### DISCUSSION TITLE
Use Create 3 and RPLiDAR for SLAM ?

### DISCUSSION THREAD

**Discussion by Yskandar** [iRobotEducation/create3_docs#54] [Questions] [answered]
Hey everyone,

I am with a robotics lab at UCLA and we are interested in this product, we're looking for a robot on which we could perform ros-based slam without struggling to much in installations and such.

To be more specific, here is what I want to do:

- **I want to mount the RPLiDAR A1 (or A2) on top of the create 3, and I want to be able to access on my computer both the topics published/subscribed by both the create 3 and the RPLiDAR. Then, I want to use the slam_toolbox (or any other package) on my computer, using the scans from the LiDAR and other informations from the create3 topics.**

After reading the docs and the issues in the discussion, the following questions remain:

1. Do I need a Raspberry Pi? I would gess so since I would like to access both the create3 and rplidar topics by wifi, since usb cables are a pain for mobile robots.

2. Is there a ROS 2 package for the create 3 that I would need to launch from my computer to access the topics of the create3 ? Or is it enough to follow the docs, meaning connect to the create 3 wifi, update the firmware and then the topics of the create3 will magically appear in my ROS2 server on my computer ?

3. The RPLiDAR's packages are in ROS1 if I am not mistaken, is this gonna be a problem? I've seen that "shamlian" used the A1 with create 3 so I'm guessing there won't be a problem.

Sorry for this long post, thank you in advance for your help !!

Yskandar

**Comment by shamlian** [comment, marked as answer]
Hi there,

First things first: don't trust that "shamlian" character.

Yes, I think for your use case, you'll need an SBC capable of running ROS 2, and with USB ports to talk to the Create 3 and the RPLIDAR. A Raspberry Pi 4 will definitely work.

The Create 3 publishes ROS 2 messages natively, and irobot_create_msgs is a ROS 2 package, so if you follow any of the compute board setup instructions on the [docs site](https://iroboteducation.github.io/create3_docs/), you should not need another package. I just also added one for a computer running [Ubuntu 20.04](https://iroboteducation.github.io/create3_docs/setup/ubuntu2004/); you can use those same directions in case you've got another OS, you'll just need to use a VM. [Here is the one for a Pi 4.](https://iroboteducation.github.io/create3_docs/setup/pi4/)

There is now an [example](https://github.com/iRobotEducation/create3_examples/tree/galactic/create3_lidar) in the create3_examples repo which includes code and mounting brackets to get this running on a Create 3 robot as described.

--Steve

**Reply by jerrynsr** [reply]
Thanks to your branch in Github iRobotEducation/create3_examples I was able to test Slamtec A1 on my Create3.  Here's a quick video.
I'm looking forward to using Cartographer and Nav2.   I included a video stream from the usb camera mounted on the bot.  For now I'm using an USB UVC camera.  I plan on buying an Oak D Lite as soon as a fixed focus version is available from a vendor in the USA.
Does the data viewed with Rviz2 appear correct ? 

Here's a link to the video: https://youtu.be/qsawrQHyU4s

**Reply by shamlian** [reply]
> Does the data viewed with Rviz2 appear correct ?

The scan does, but you should also be able to see an occupancy grid if you enable it, which might help you localize your brain better to the laser scan. Edit: looks like your physical and virtual sensor location didn't match.

**Reply by shamlian** [reply]
@justinIRBT checked in an rviz config file in the branch, which I've independently verified to work. If you are using CycloneDDS, you might have problems with multi-interface on both your computer and your SBC (this happened to me). You can either set up XML profiles as recommended in our [network config](https://iroboteducation.github.io/create3_docs/setup/xml-config/) page, or be lazy and use FastDDS. No judgement. ;-)

**Comment by Yskandar** [comment]
Hi Steven,

Thank you for answering this quickly, I'll make sure not to trust him do not worry.

Looking forward to the step-by-step tutorial !

Yskandar

**Comment by jerrynsr** [comment]
I’ve pulled down the update and already using it :)  I am using rmw_fastrps_cpp.
I am getting a warning in Rviz, no transform for base_footprint and laser_frame.  Is that to be expected or did I miss something ?

**Reply by shamlian** [reply]
Should "just work;" I just tried it with my Pi, laser, and robot, with my Ubuntu 20.04 machine. Sensors_launch.py creates the transform. Are all three scripts running on the Pi? And have you ensured that the RMW is set the same on the Create, the Pi, and your computer?

**Reply by ajjaya18** [reply]
Hi, I have the same error, no transform for base_footprint and laser_frame, how to solve this? Help me please.

**Comment by jerrynsr** [comment]
All three scripts were running.   I cleaned out my workspace,  did a fresh git checkout.  I am running colcon build now.

**Comment by jerrynsr** [comment]
Hi Steve,
      I attached a photo of the results I have so far.  The orientation
of the map looks wrong.  I mounted the LIDAR with the scanning motor
facing the back end of the Create3.  Opposite from whatyou have in your
photos.   Do I need to modify the sensors_launch.py or
slam_toolbox_launch.py file ?   
Thanks!

**Reply by shamlian** [reply]
Yes, as I mentioned in passing, the tf transform is set up in sensors_launch.py. If you've got the sensor in a different geometric arrangement relative to the motion of the robot than the SLAM solver is trying to solve for, it's going to have a very hard time. If you've rotated the LIDAR by 180 degrees, it's going to have an impossible time. Either use our arrangement (center of LIDAR is 12 mm behind the center of the robot, mounted as shown) or modify the parameters in the static transform publisher launched from sensors_launch.py.

**Comment by jerrynsr** [comment]
The map would make more sense to me once PointCloud data is displayed
in RVIZ2.  Is the next step to get the PointCloud Publisher running ?

**Reply by shamlian** [reply]
There's no point cloud in this example -- we're using a planar LIDAR

**Comment by jerrynsr** [comment]
Is there going to be a mapping example ?  Also I'm using the gazebo
simulator examples to get an idea of how to use the ir sensors for
obstacle avoidance and add that to wall_follower and coverage.I would
like t use both the LIDAR and IR sensors for obstacle avoidance.  Just
in case there's something on the floor the LIDAR does not pick up.
Is there going to be a point cloud example ?

**Reply by shamlian** [reply]
I can't give you a schedule, but we do have aspirations for both. FWIW this example will generate a map once you fix your kinematics.

**Comment by jerrynsr** [comment]
I wish the ROS2 documentation was a good as ROS1.  
What should I use here to rotate the LIDAR 180 degrees ? return
LaunchDescription([

        Node(package='tf2_ros',
executable='static_transform_publisher',

            arguments=['-0.012', '0', '0.144', '0', '0', '0',
'base_footprint', 'laser_frame']),
Can you send me a link regarding ROS2 static_transform_publisher
arguments ?    I found examples, but nothing describing all the
parameters that can be passed.

**Reply by shamlian** [reply]
it's X, Y, Z, roll, pitch. yaw, IIRC, so just change that last '0' to '3.1416'

**Comment by jerrynsr** [comment]
Would this work ? Do I need to change the offset as well ?
!/usr/bin/env python3

# Copyright 2022 iRobot Corporation. All Rights Reserved.



from launch import LaunchContext

from launch import LaunchDescription

from launch_ros.actions import Node

from ament_index_python.packages import get_package_share_directory

import math



def generate_launch_description():

    return LaunchDescription([

        Node(package='tf2_ros',
executable='static_transform_publisher',

            arguments=['-0.012', '0', '0.144', '3.14159', '0', '0',
'base_footprint', 'laser_frame']),

        Node(

            package='rplidar_ros',

            parameters=[

                get_package_share_directory("create3_lidar") +
'/config/rplidar_node.yaml'

                ],

            executable='rplidar_composition'

        ),

    ])

~

**Reply by shamlian** [reply]
I would highly suggest you go through the official [tf2 tutorials](https://wiki.ros.org/tf2/Tutorials). What you have done is rotate the laser on the X axis, which I don't think is what you meant to do (unless you mount the laser upside-down). You will need to make the tf tree accurate to wherever it is you've stuck your laser rangefinder. The values supplied are for the location that we have shown it mounted in the example, which is at -12 mm on the X axis (origin in the center of rotation), 144 mm above the ground, with no rotation, as defined by [SLAMTEC's documentation](https://github.com/robopeak/rplidar_ros/wiki).

**Comment by jerrynsr** [comment]
ok.. that makes sense.   I'll mount the lidar with the motor facing the
front.

**Comment by jerrynsr** [comment]
Success!

First pass.  The RPI4 crashed, but the EXT3 journal saved the day.    It looks lit it will take several passes to get a solid map.  This looks accurate.
